src/Bicep.Core/Semantics/ObjectParser.cs

```diff
@@ -0,0 +1,79 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using Bicep.Core.Diagnostics;
+using Bicep.Core.Parsing;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Bicep.Core.Semantics
+{
+    public abstract class ObjectParser : IObjectParser
+    {
+        public bool TryExtractFromObject(string fileContent, string? tokenSelectorPath, IPositionable[] positionable, [NotNullWhen(false)] out ErrorDiagnostic? errorDiagnostic, [NotNullWhen(true)] out JToken? newToken)
+        {
+            errorDiagnostic = null;
+            newToken = this.ExtractTokenFromObject(fileContent);
+            if (newToken is not { })
+            {
+                // Instead of catching and returning the YML parse exception, we simply return a generic error.
```

Comment:
```suggestion
                // Instead of catching and returning the parsing exception, we simply return a generic error.
```
[nit]: in this context, what's being parsed may be JSON or YAML